src/combinatorics/rado/rank_function.lean

```diff
@@ -0,0 +1,224 @@
+/-
+Copyright (c) 2023 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+
+import combinatorics.rado.auxiliary
+
+/-!
+# Rank functions
+
+We define the type of rank functions on a type `α` and prove some properties.
+
+See the docstring of `rank_fn` for the definition.
+-/
+
+section rk_fn_def
+
+variables (α : Type*) [decidable_eq α]
+
+/-- A *rank function* on a type `α` is a function `r` from `finset α` to natural numbers such that:
+* `r ∅ = 0`
+* `r s ≤ r t` if `s ⊆ t`
+* `r ({a} ∪ s) ≤ r s + 1`,
+* `r ({a} ∪ s) = r s` and `r ({b} ∪ s) = r s` imply `r ({a, b} ∪ s) = r s`
+
+(This definition is equivalent to the definition one can find on the
+(Wikipedia page for matroids)[https://en.wikipedia.org/wiki/Matroid].
+But note that we do not assume that the base type `α` is finite.)
+-/
+@[ext]
+structure rank_fn :=
+(to_fun : finset α → ℕ)
+(empty' : to_fun ∅ = 0)
+(mono' : monotone to_fun)
+(insert_le' : ∀ a s, to_fun (insert a s) ≤ to_fun s + 1)
+(dep' : ∀ {s a b}, to_fun (insert a s) = to_fun s → to_fun (insert b s) = to_fun s →
+          to_fun (insert b (insert a s)) = to_fun s)
```

Comment:
You can drop the `decidable_eq` assumption if you instead use
```suggestion
(insert_le' : ∀ a s ha, to_fun (cons a s ha) ≤ to_fun s + 1)
(dep' : ∀ {s a b} ha hb (hab : a \ne b), to_fun (cons a s ha) = to_fun s → to_fun (cons b s hb) = to_fun s →
          to_fun (cons b (cons a s hs) sorry) = to_fun s)
```
where that `sorry` should be easy to prove.

Reply:
(you can then prove the `insert` versions by a case bash)

Reply:
Is removing the `decidable_eq` assumption really worth it? You can't do a lot with `finset`s without it anyway...